core/src/main/java/org/infinispan/commands/functional/ReadWriteKeyValueCommand.java

```diff
@@ -103,7 +103,9 @@ public Object perform(InvocationContext ctx) throws Throwable {
       // If the value has been update while on the retry, use the newer value.
       // Also take into account that the value might have been removed.
       // TODO: Configure equivalence function
-      if (valueUnchanged(e, prevValue, value) || valueRemoved(e, prevValue)) {
+      // TODO: this won't work properly until we store if the command was executed or not...
+      Object oldPrevValue = e.getValue(hasFlag(Flag.COMMAND_RETRY));
```

Comment:
With `ReadCommittedEntry` (as so far the func commands are non-tx only) this will always return the current value. As the entry is removed from context in `EntryWrappingInterceptor.removeFromContextOnRetry` before wrapping the value again on a new command execution, it returns always the value currently in DC, and that's the best you can get.
I admit that the `getValue(retry)` may be unnecessary (I could try to remove that logic and run testsuite again), this code went through several iterations and given that it's broken by design from the beginning...